File: 7-evaluation/lib/expression.rb
require 'scanner'

module Expression
  def self.define_expression_type(name, *fields)
    Struct.new(*fields) do
      define_method :accept do |visitor|
        visitor.public_send(:"visit_#{name.downcase}", self)
      end
    end
  end

  Binary = define_expression_type('Binary', :left, :operator, :right)
  Grouping = define_expression_type('Grouping', :expression)
  Literal = define_expression_type('Literal', :value)
  Unary = define_expression_type('Unary', :operator, :right)
end

File: rust-lox/src/expression.rs
use super::scanner;

#[derive(Debug, PartialEq)]
pub enum Expression<'a> {
  Binary {
    left: Box<Expression<'a>>,
    operator: scanner::Token<'a>,
    right: Box<Expression<'a>>,
  },
  Grouping {
    expression: Box<Expression<'a>>,
  },
  Literal {
    value: f32,
  },
  Unary {
    operator: scanner::Token<'a>,
    right: Box<Expression<'a>>,
  },
}

#[cfg(test)]
mod tests {
  use super::*;

  #[test]
  fn test_something() {
    let left = Box::new(Expression::Literal { value: 666.0 });
    let right = Box::new(Expression::Literal { value: 42.0 });
    Expression::Binary {
      left: left,
      operator: scanner::Token {
        token_type: scanner::TokenType::Plus,
        text: "+",
        line: 1,
      },
      right: right,
    };
  }
}
